### PR TITLE
Feat/matching service frontend

### PR DESCRIPTION
**Pull request by Aloynz**
Things done/fixed:
- Cancel Matching Logic
- Both client sides successfully display their match when match is found between them

Existing bug 1:
- somehow after a user cancel his matching, the client is disconnected but the request is still in the queue and being processed. To replicate:
1) Press match on client1
2) Click Cancel matching on client1 and immediately on client2, click match
Expected outcome: No match should be found since client1 already cancelled matching
Actual outcome: A match between client1 and client2 

What is printed on console:
```
New client connected: 70_WLB8FVAgE-d4bAAAB (client1 connected after pressing match)
No immediate match for Cedric, waiting for 30 seconds. (no immediate result so start 30s timer)
User Cedric joined their room for match updates. (client1 join room and wait 30s)
Client disconnected: 70_WLB8FVAgE-d4bAAAB (client1 press cancel matching and disconnect)
New client connected: 8z4q1pBIcH-UkfUyAAAD (client2 connected after pressing match)
Match found between admin and Cedric (the issue!)
User admin joined their room for match updates. (this should be correct since there is no match as client1 is has cancelled matching, but it is instead showing the match on client2's side)
Client disconnected: 8z4q1pBIcH-UkfUyAAAD 
```

Existing bug2:
- after client disconnect, the backend is still checking for match
To replicate:
- Click Match in client, then cancel before timer runs
```
What is printed on console:
Client disconnected: Li_qA57XbIh6QHtvAAAF
No match found for Cedric after 30 seconds. (shouldnt have this as the user already cancelled his request)
```

### CODE REVIEW

**Comment by CedricChia123** [comment]
LGTM, match reflects for both users. The bugs are reproducible, will follow up

**Comment by chiewyx** [comment]
- implemented fix for matching bug by ensuring users join room before proceeding to find a match
- implemented fix for match cancellation bugs
  - logic was correct but frontend was sending userId as request, and therefore redis key could not be found

**Comment by JunHao247** [comment]
> * implemented fix for matching bug by ensuring users join room before proceeding to find a match
> * implemented fix for match cancellation bugs
>   
>   * logic was correct but frontend was sending userId as request, and therefore redis key could not be found

Tested, output works as expected. LGTM

**Review by JunHao247 — APPROVED** [review]
LGTM, both users are matched

**Comment by CedricChia123** [comment]
> * implemented fix for matching bug by ensuring users join room before proceeding to find a match
> * implemented fix for match cancellation bugs
>   
>   * logic was correct but frontend was sending userId as request, and therefore redis key could not be found

tested as well, LGTM.

File: matching-service/src/service/match-service.ts [hunk 1]
```diff
@@ -2,22 +2,17 @@ import { Server } from "socket.io";
 import { MatchRequest } from "../types/match-types";
 import { connectRabbitMQ, getChannel } from "../queue/rabbitmq";
 import Redis from "ioredis";
-
-let matchQueue: MatchRequest[] = [];
+import { activeMatches } from "../server";
```
Comment (chiewyx) [inline matching-service/src/service/match-service.ts:5]: can help to clean up this

File: matching-service/src/server.ts [hunk 2]
```diff
@@ -36,12 +45,21 @@ server.listen(PORT, async () => {
...
+    // Send acknowledgment back to the client
+    callback({ success: true });
   });
 
   socket.on("disconnect", () => {
     console.log("Client disconnected:", socket.id);
   });
 });
+
+export { activeMatches };
```
Comment (chiewyx) [inline matching-service/src/server.ts:65]: this as well

File: matching-service/src/server.ts [hunk 3]
```diff
@@ -20,7 +20,16 @@ const io = new Server(server, {
...
+// Middleware to attach the Socket.io instance to the req object
+app.use((req, res, next) => {
+  req.io = io;
+  next();
+});
+
+app.use('/api', matchRoutes);
+
+// In-memory store for matches
+const activeMatches = new Map();
```
Comment (chiewyx) [inline matching-service/src/server.ts:32]: this also

**Review by chiewyx — COMMENTED** [review]
I think just need to clean up the use of activeMatches map (i think not using alrd but we r still exporting and importing it)

**Comment by Aloynz** [comment]
- cleanup and remove activematches
- update readme for cancel match api endpoint

**Comment by chiewyx** [comment]
LGTM, will merge

**Pull request merged**